sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1908,7 +1928,13 @@ class Dataset[T] private[sql](
   *         }
   *       }
   *     }
+  *     def onQueryStarted(event: QueryStartedEvent): Unit = {}
+  *     def onQueryTerminated(event: QueryTerminatedEvent): Unit = {}
   *   })
+  *
+  *   // Observe row count (rc) and error row count (erc) in the streaming Dataset
+  *   val observed_ds = ds.observe("my_event", count(lit(1)).as("rc"), count($"error").as("erc"))
+  *   observed_ds.writeStream.format("...").start()
   * }}}
   *
   * @group typedrel
```

Comment:
Can we add `@Unstable` or `@DeveloperApi` while we're here? The usage with `QueryExecutionListener` should be considered as one of both.